### ISSUE TITLE
Fluent Assertion Should().Be() returns false positive

### ISSUE THREAD

**Issue by onlyforkbj** [fluentassertions/fluentassertions#569]
Our Team uses FluentAssertions alongside XUnit for unit testing. In the given code below, we are trying to test the Status Code 400 (Bad Request) from our controllers by Passing null to one of our argument to the Controller method. 

Unit Test:

[Fact]
public async Task PartiallyUpdatePerson_InvalidJsonPatchDocument_ReturnsStatusCode400()
{
     var id = new Guid("47217150-1eff-4c7e-84a3-2b57001a18d6");

     var result = await sut.PartiallyUpdatePerson(id, null) as ObjectResult;

      result?.StatusCode.Should().Be(400);
}

SUT:

[HttpPatch("{id}", Name = "PartiallyUpdatePerson")]
public async Task<IActionResult> PartiallyUpdatePerson(Guid id, [FromBody] JsonPatchDocument<PersonPatchDto> personToPatch)
{
     if (personToPatch == null || id == Guid.Empty)
     {
           return BadRequest();
     }
     .
     .
     .
     .
     return NoContent();
}
This unit test passes and everyone is happy with it. 

The problem is when we try to change the Should().Be() assertion value. If we change from 400 to anything other than that, we expect the Test to fail. But it still passes with Statuscode <> 400. 

**To Summarize this is what is happening now:**

result?.StatusCode.Should().Be(400)  --> Test Run Passes
result?.StatusCode.Should().Be(401)  --> Test Run Still Passes where it should have failed

On the other hand, if we use the regular assertion, 

Assert.True(result?.StatusCode == 400) --> Test Run Passes 
Assert.True(result?.StatusCode == 401) --> Test Run Fails 

I'm not sure if we're doing anything wrong or if there is an issue with Should().Be(). Please let me know what you think about it. 

Thanks,
Balaji

**Comment by dennisdoomen** [comment]
Considering the basic syntax, that is really weird. What is the type of the class om which `Be` lives, according to the compiler?

**Comment by onlyforkbj** [comment]
If I understand your quesiton correctly, the controller action returns a BadRequestResult as ObjectResult and upon which ´Be()´ is used.

**Comment by dennisdoomen** [comment]
But what class does `Should()` return here? It helps me understand the path FA is following.

**Comment by SaroTasciyan** [comment]
I don't know the type of `result` variable there but that _null conditional operator_ is kind of suspicious to me. It might be that the result is actually `null`, therefore the entire expression evaluates to a `null` without having the chance to call `Should().Be(...)` .

I did a bit of testing as follows:
```c#
Result result400 = new Result { StatusCode = 400 };
Result result401 = new Result { StatusCode = 401 };
Result nullResult = null;

result400?.StatusCode.Should().Be(400);
result401?.StatusCode.Should().Be(401);
nullResult?.StatusCode.Should().Be(400);
nullResult?.StatusCode.Should().Be(401);

// Fails as expected here with message: Expected value to be 401, but found 400
result400?.StatusCode.Should().Be(401);
```

Using null conditional/elvis, you might expect it wouldn't fail with `nullResult` assertions.

**Comment by krajek** [comment]
@dennisdoomen you added `requires-clarification` tag, but the issue is clear to me.
@SaroTasciyan explained, that the problem was with the usage of operator .? which does not call right-hand side of expression at all, in case of `null` subject.

```
int? subject = null;
subject?.Should().Be(400);
```

Above code does not throw.
FA itself cannot help here. I would suggest closing the issue here.
I will check if static analyzer for FA catches this one, and report an issue in their repository if it does not.

**Comment by dennisdoomen** [comment]
I think we should wait for @onlyforkbj to confirm that this was indeed the problem.

**Comment by jnyrup** [comment]
Closing due to inactivity and several good descriptions/explanations of what might have caused the problem.

**Comment by NamrataKankaria** [comment]
I am also facing the same issue. 
When i use try and catch block and under my try block when I use fluent assertions. My test gets passed even though it should fail.

Sample code -

[Test]
        public void TestingFuent()
        {
            try
            {
                "HI".Should().Be("as");
            }
            catch (Exception ex)
            {

            }
        }

-- The above code could will go to catch block. But Test will be shown as PASSED

**Comment by dennisdoomen** [comment]
> The above code could will go to catch block. But Test will be shown as PASSED

But that's because you're swallowing the exception that FA throws

**Comment by NamrataKankaria** [comment]
> > The above code could will go to catch block. But Test will be shown as PASSED
> 
> But that's because you're swallowing the exception that FA throws

Hi Dennis,
How do I resolve this? 
I am required to put try and catch block in my code for other purpose.

**Comment by dennisdoomen** [comment]
No. Why would you need a `try/catch` block at all?

**Comment by NamrataKankaria** [comment]
> No. Why would you need a `try/catch` block at all?

For statements other than assert statements. 
Like trying to access an element and do some operation etc and later using fluent assertion statement to verify our test
